design/single-pod-helm-chart.md

```diff
@@ -0,0 +1,146 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Single-pod Helm chart
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [Title](#title)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories [optional]](#user-stories-optional)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+Provide a Helm chart to deploy Metal3 and Ironic as a single
+pod in a Kubernetes cluster.
+
+## Motivation
+
+### Goals
+
+The goal is to support a popular way to deploy Kubernetes applications
+to simplify creation of development environments on top of arbitrary
+Kubernetes clusters.
+
+Another goal to preare to set a standard for production-grade deployment
+of Metal3 and its components.
+
+### Non-Goals
+
+Providing end-to-end bootstrap sequence for Metal3 and Ironic is not a
+a goal of this design.
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user of Metal3, I want to install it in my existing Kubernetes
+cluster using Helm.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Initial implementation includes 2 Helm charts that create corresponding
+pods in a Kubernetes cluster:
+
+* Ironic pod
+* Metal3 pod
+
+The charts shall be added as a separate repository in metal3-io space.
+
+### Risks and Mitigations
+
+None.
+
+## Design Details
+
+Helm charts will require a separate repository (metal3-helm-charts)
+to be created.
+
+In future, a CI environment that will build and test the charts,
+will have to be created.
+
+Potentially, the Helm charts will require changes in the way the
+Docker images for components of Ironic and Metal3 are built. The
+changes will include additional parameters that will be exposed
+through the charts metadata (values.yaml files).
+
+### Work Items
+
+1. Create a Helm chart for Ironic and its components:
+   * ``ironic``
```

Comment:
Ideally, you should split ironic-api and ironic-conductor since running two processes in one container is not the best practice.

Reply:
Agreed.   I think this is also an opportunity to standardize the logging practices across all single-process pods as another example of proper production behavior.  All pods should be logging to stdout/stderr.

Reply:
It occurs to me that this point may not have been captured very well in any design docs, but a design goal we had discussed before building this was that we really wanted the baremetal management component to be ephemeral worker pods someday that we could scale out as necessary.

First we just need a self contained Ironic that we know can work without any real persistent storage.  The baremetal-operator needs to be able to re-populate Ironic with information as needed at any point.

Eventually thought, it would be nice to be able to spin up bare metal management pods to perform management tasks as needed.  That's a ways off probably ... I like the idea of just a single unit while we make things work at all.